apis/shared-dimensions/lib/domain/shared-dimensions.ts

```diff
@@ -36,42 +43,52 @@ interface GetSharedTerms {
   validThrough?: Date
 }
 
-export function getSharedTerms({ sharedDimensions, freetextQuery, validThrough, limit = 10, offset = 0 }: GetSharedTerms) {
-  const term = $rdf.variable('term')
-  const name = $rdf.variable('name')
-  const sharedDimension = $rdf.variable('sharedDimension')
+export async function getSharedTerms<C extends StreamClient | ParsingClient>({ sharedDimensions, freetextQuery, validThrough, limit = 10, offset = 0 }: GetSharedTerms, client: C): Promise<C extends StreamClient ? Stream : Quad[]> {
+  const shape = await loadShape()
+  if (!isGraphPointer(shape)) {
+    throw new Error('Multiple shapes found')
+  }
 
-  let select = SELECT.DISTINCT`${term}`
-    .WHERE`
-      ${sharedDimension} a ${meta.SharedDimension} .
-      VALUES ${sharedDimension} { ${sharedDimensions} }
-      ${term} ${schema.inDefinedTermSet} ${sharedDimension} .
-      ${term} ${schema.name} ${name} .
-    `
+  shape.any().has(sh.limit).deleteOut(sh.limit).addOut(sh.limit, limit)
+  shape.any().has(sh.offset).deleteOut(sh.offset).addOut(sh.offset, offset)
+
+  const filterShape = shape.any().has(sh.filterShape).out(sh.filterShape)
+
+  const dynamicProperties = await getDynamicProperties(sharedDimensions)
+  dynamicProperties.forEach(predicate => {
+    shape.out(sh.property).out(sh.node).addOut(sh.property, propertyShape => {
+      propertyShape.addOut(sh.path, predicate)
+    })
+  })
+
+  filterShape
+    .out(sh.property)
+    .has(sh.path, schema.inDefinedTermSet)
+    .addList(sh.in, sharedDimensions)
 
   if (freetextQuery) {
-    select = select.WHERE`${textSearch(term, schema.name, freetextQuery)}`
+    filterShape
+      .out(sh.property)
+      .has(sh.path, schema.name)
+      .addOut(hydra.freetextQuery, freetextQuery)
   }
 
   if (validThrough) {
-    select = select.WHERE`OPTIONAL {
-      ${term} ${schema.validThrough} ?validThrough .
-    }
-
-    FILTER (
-      !bound(?validThrough) || ?validThrough >= ${toRdf(validThrough)}
-    )`
+    const placeholder = $rdf.literal('VALID-THROUGH')
+    filterShape.out(sh.expression).deleteOut(sh.deactivated)
+    filterShape.dataset.match(null, null, placeholder).forEach((quad) => {
+      filterShape.dataset.delete(quad).add($rdf.quad(quad.subject, quad.predicate, toRdf(validThrough), quad.graph))
+    })
   }
 
-  return DESCRIBE`
```

Comment:
I changed a DESCRIBE query to a CONSTRUCT generated from SHACL shape using https://shape-to-query.hypermedia.app/docs/